### ISSUE TITLE
Cache invalidation following repair / streaming works on sstable token ranges - clearing at times (too) large parts of the cache

### ISSUE THREAD

**Issue by slivne** [scylladb/scylladb#9136]
copy pasting from internal thread

Avi  suggested

> What can we do?
> 
> 1. merge the new sstables into cache (like we merge memtables into cache). More I/O, but if repairs usually generate small amounts of data it's not too bad.
> 2. invalidate by partition. Walk the cache partitions, check if the partition exists in the repair sstables using bloom filters, invalidate if there is a hit. This will work well except for large partitions.
> 3. combination, select 1 or 2 depending on the amount of data (for example for repair based node operations we don't want to merge)
> 

Tomek suggested

> The md sstable format has now clustering range annotation. We could use that to narrow down invalidation range within partitions.

There are other suggestions still need to a line on a solution

**Comment by michoecho** [comment]
@avikivity @asias
We talked on the last latency tracker about decommission invalidating all row caches in the cluster.
I guess this is the issue that tracks the problem.

**Comment by dorlaor** [comment]
Btw, when we stream/repair, do we bypass the cache, so we wouldn't pollute
it with a one-time
used objects

**Comment by michoecho** [comment]
> Btw, when we stream/repair, do we bypass the cache, so we wouldn't pollute it with a one-time used objects

Yes, streaming is done directly to sstables.

**Comment by raphaelsc** [comment]
> Btw, when we stream/repair, do we bypass the cache, so we wouldn't pollute it with a one-time used objects
> […](#)

streaming/repair producer operates on sstable set, so it's not affecting cache in any way. the receiver invalidates cache on newly added files, therefore it's the one suffering with cache invalidation.

Tomek's suggestion on narrowing down clustering range, using position metadata present in Statistics.db, would greatly help with large partitions.

Row cache invalidation would become an iteration on the sstable's key range, and the cache would check sstable bloom filter for each partition present in it, and then narrow down the row range from sstable's position metadata.

**Comment by raphaelsc** [comment]
Position metadata, however, is only limited to min and max values, therefore, it's not available per partition. Perhaps we could involve the index, comes with I/O cost, but that will have good effect on read latency down the road.

**Comment by dorlaor** [comment]
This is an important improvement, since from time to time customers have
unexplained issues when
running admin operations and we like to minimize these effects. Thanks for
the full description Michal!

**Comment by asias** [comment]
> The description in the issue is extremely unhelpful, so to clarify/remind the problem:
> 
> The row cache in Scylla is a partial, but exact, in-RAM image of the node's sstable set. When a new sstable is added, the row cache must be updated to recover its exactness — either the contents of the sstable must be merged into the cache, or parts of the cache must be removed in a way that ensures that nothing in the cache is overwritten by the sstable.
> 
> Currently, when an sstable is added, we just invalidate the entire range of the sstable — from the least to the greatest token it contains. This is wasteful, because it removes from the cache many entries which are not overwritten by the sstable, and are still up-to-date.
> 
> Streaming works by writing directly to sstables. Therefore, for example, if ranges [1, 10] and [1001, 1010] are streamed, the entire [1, 1010] range will be invalidated on the receiver, even though 99% of it was unaffected, and is still up-to-date.
> 
> The case of decommission is particularly bad, because it streams many small ranges randomly scattered over the ring — if you take several hundred random 64-bit integers, their total range will cover almost the entire 64-bit space. In other words, decommission invalidates ~99% of the entire row cache on all nodes in the cluster, potentially wreaking havoc on performance until the cache recovers (which will take several minutes).
> 
> For decommission, the fix is simple — just tell the receiver which ranges were streamed, so it can invalidate only the individual streamed ranges, and not their total span. (And since a decommission receiver shouldn't own/replicate these ranges before decommission, this invalidation will be a no-op — so this is enough to fully fix the wasteful invalidation for the special case of streaming during decommission).
> 
> More general fixes (i.e. for repair and for invalidation on partition and sub-partition level) are more involved, and were discussed in [scylladb/scylla-enterprise#1720](https://github.com/scylladb/scylla-enterprise/issues/1720) and by @raphaelsc above.
> 
> My point here is: perhaps the case of decommission is worth fixing by itself, without considering the general problem of repair, since fixing the former is relatively easy.

Hello @michoecho, thanks for picking this up. I want to add some points, hoping it can help. 

1) The classic streaming for decommission will generate sstables covers multiple vnode ranges. I think this is the  reason why you saw lots of ranges are invalidated. However, with RBNO decommission will generate sstalbes covering at most one vnode range. You can play with RBNO decommission to check it the cache behavior. 

2) A node might re-own the range it owned in the past. For example, n1 is in the cluster,  n2 is added, n1 loses some range, then n2 is decommissioned, n1 re-own some ranges.

**Comment by avikivity** [comment]
> 
> 1. The classic streaming for decommission will generate sstables covers multiple vnode ranges. I think this is the  reason why you saw lots of ranges are invalidated. However, with RBNO decommission will generate sstalbes covering at most one vnode range. You can play with RBNO decommission to check it the cache behavior.

Won't offstrategy consolidate them to one sstable?

**Comment by asias** [comment]
> > 1. The classic streaming for decommission will generate sstables covers multiple vnode ranges. I think this is the  reason why you saw lots of ranges are invalidated. However, with RBNO decommission will generate sstalbes covering at most one vnode range. You can play with RBNO decommission to check it the cache behavior.
> 
> Won't offstrategy consolidate them to one sstable?

As soon as the sstables are added even to the maintenance set, the cache is invalidated.  The sstables in maintenance set are included to to serve read.

**Comment by raphaelsc** [comment]
If we think it's better to invalidate cache once, which I think it's reasonable assumption as we'll have to go through invalidation process only once, we could theoretically read from maintenance set like we read from memtables (in front of cache). Of course still using the incremental reader. On off-strategy compaction completion, then we proceed to invalidate cache on sstable(s) added to main set.

Another good reason is that all new sstables in maintenance set aren't generated once, so we're having multiple invalidations at an interval (which is the amount of time for repair / streaming to produce a new sstable).

note that the off-strategy output may be more than 1 sstable, depending on the compaction strategy used. TWCS off-strategy will produce 1 sstable per window.

**Comment by raphaelsc** [comment]
But of course, off-strategy generating more than 1 sstable is not a blocker for switching to a single invalidation per off-strategy.

**Comment by raphaelsc** [comment]
A disadvantage that comes to mind is that once we consolite all data into a single sstable, that sstable may end up with a coarse-grained token range. So it would be important to have this issue fixed first before we do anything in this area.

**Comment by asias** [comment]
> A disadvantage that comes to mind is that once we consolite all data into a single sstable, that sstable may end up with a coarse-grained token range. So it would be important to have this issue fixed first before we do anything in this area.

@raphaelsc Avi is not suggesting we should consolidate all data into a single sstable and then add the sstable and invalidate cache once. Bigger sstable means bigger token ranges with bigger gaps in it, which is bad for cache invalidation.

**Comment by raphaelsc** [comment]
> > A disadvantage that comes to mind is that once we consolite all data into a single sstable, that sstable may end up with a coarse-grained token range. So it would be important to have this issue fixed first before we do anything in this area.
> 
> @raphaelsc Avi is not suggesting we should consolidate all data into a single sstable and then add the sstable and invalidate cache once. Bigger sstable means bigger token ranges with bigger gaps in it, which is bad for cache invalidation.

Which is **exactly** what my comment above says. If we were to invalidate cache only once for any reason, I said this issue must be fixed first. My last comment dismissed my first two comments. Invalidating once per vnode range is the best we have right now, but note it can still suffer by the same problem. A sstable generated for a vnode range X can still have terrible density, meaning many tokens will be wiped off the cache for no reason. So invalidating once per vnode range is still far from being optimal. And that's why this issue was opened in the first place.

**Comment by asias** [comment]
@michoecho @avikivity 

Given

1) A list of token ranges for all the caches in the system

2) A list of tokens ranges the node owns

3) The token range of the sstable being added

Can we intersect 1), 2) and 3), and invalidate only the the intersected range?

**Comment by michoecho** [comment]
> @michoecho @avikivity
> 
> Given
> 
>     1. A list of token ranges for all the caches in the system
> 
>     2. A list of tokens ranges the node owns
> 
>     3. The token range of the sstable being added
> 
> 
> Can we intersect 1), 2) and 3), and invalidate only the the intersected range?

The cache is assumed to reflect the SSTable set exactly. Your proposal keeps this property only if it is guaranteed that the sstable being added only contains tokens which the node owns. (If it contains a token which the node doesn't own, then it will be added to the SSTable set, but not invalidated in the cache, destroying the exactness). Is it guaranteed?

Anyway, I think this idea is going into dangerous territory — adding "token ownership" makes a local problem (exactly caching local SSTables) into a distributed problem (exactly caching the subset of SSTables which might possibly be read).

**Comment by asias** [comment]
> > @michoecho @avikivity
> > Given
> > ```
> > 1. A list of token ranges for all the caches in the system
> > 
> > 2. A list of tokens ranges the node owns
> > 
> > 3. The token range of the sstable being added
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can we intersect 1), 2) and 3), and invalidate only the the intersected range?
> 
> The cache is assumed to reflect the SSTable set exactly. Your proposal keeps this property only if it is guaranteed that the sstable being added only contains tokens which the node owns. (If it contains a token which the node doesn't own, then it will be added to the SSTable set, but not invalidated in the cache, destroying the exactness). Is it guaranteed?

I forgot to mention that by 2), I meant we first invalidate the cache that does not belong to the node using 1) and 2). Then we do the intersect to figure out what is exactly needs to be invalidated for the sstable we add. 

> 
> Anyway, I think this idea is going into dangerous territory — adding "token ownership" here converts a local problem (exactly caching local SSTables) into a distributed problem (exactly caching the subset of SSTables which might possibly be read).

I understand the concern, however we already invalidate cache using the tokne ranges a node owns in nodetool cleanup. Since we only invalidate the chache, it is not going to be that scaring. In worse case, we lose the cache and read from disk. No correctness issue.

**Comment by michoecho** [comment]
> I understand the concern, however we already invalidate cache using the tokne ranges a node owns in nodetool cleanup.

Wait, let's be more precise about what "the node" means, because I'm not sure which nodes we are talking about. I understand that your idea is to invalidate on the sstable _receiver_ the intersection of the arriving sstable's range with the token ranges of the sstable _sender_?

In the case of cleanup, for the exactness property of cache, token ownership doesn't matter at all — during cleanup we filter both the cache and the sstables through the same token ranges, so the exactness will be preserved no matter what these ranges are. So there is no distributed metadata coming into the picture.

I don't understand the use cases of streaming well enough (that is, I'm not familar with Scylla's topology change procedures) to discuss correctness here, but it seems to me that the intersection idea (which I guess is a way to avoid extending the streaming RPC protocol with explicit metadata of streamed ranges) only works if it is guaranteed that the receiver and the sender are in agreement about which ranges the sender owns (this is where the distributed problem is), and either the sender doesn't send anything outside of that range, or the receiver filters put anything outside of the range.
If that's guaranteed, then there's no problem.

**Comment by asias** [comment]
> > I understand the concern, however we already invalidate cache using the tokne ranges a node owns in nodetool cleanup.
> 
> Wait, let's be more precise about what "the node" means, because I'm not sure which nodes we are talking about. I understand that your idea is to invalidate on the sstable _receiver_ the intersection of the arriving sstable's range with the token ranges of the sstable _sender_?

The node is the receiving node only. We do not care about token ranges of the sender.

> 
> In the case of cleanup, for the exactness property of cache, token ownership doesn't matter at all — during cleanup we filter both the cache and the sstables through the same token ranges, so the exactness will be preserved no matter what these ranges are. So there is no distributed metadata coming into the picture.

By the same token ranges, you meant the token ranges owned by the node in question or something else? 

> 
> I don't understand the use cases of streaming well enough (that is, I'm not familar with Scylla's topology change procedures) to discuss correctness here, but it seems to me that the intersection idea (which I guess is a way to avoid extending the streaming RPC protocol with explicit metadata of streamed ranges) only works if it is guaranteed that the receiver and the sender are in agreement about which ranges the sender owns (this is where the distributed problem is), and either the sender doesn't send anything outside of that range, or the receiver filters out everything outside of the range. If that's guaranteed, then there's no problem.

The intersection idea is about to invalidate only the relevant ranges no matter the start and end of sstable. The topology change procedures and token ownship change are not relevant.

**Comment by michoecho** [comment]
> By the same token ranges, you meant the token ranges owned by the node in question or something else?

Yes, I meant the token ranges owned by the node performing cleanup.

> The node is the receiving node only. We do not care about token ranges of the sender.

Oh. In that case I don't understand the idea at all.

**Comment by asias** [comment]
Consider:

cache [1, 10] [20, 30] [50,100]
owed token range [1, 8] [22, 25] [50, 80]
sstable range [1, 1000], which only has 3 tokens 1 , 70, 1000

with intersection idea, we first drop the cache so that we have the new cache

cache [1, 8] [22, 25], [50, 80]

Then we check the sstable against the trimed cache.

range [1, 8] hits tokens within the sstable, we invalidate cache [1, 8].
range [22, 25] does not hit anyt oken within the sstable, we skip to invalidate cache [22, 25]
range [50, 80] hit tokens within sstable, we skip to invalidate cache [50, 80]

**Comment by michoecho** [comment]
> Consider:
> 
> cache [1, 10] [20, 30] [50,100] owed token range [1, 8] [22, 25] [50, 80] sstable range [1, 1000], which only has 3 tokens 1 , 70, 1000
> 
> with intersection idea, we first drop the cache so that we have the new cache
> 
> cache [1, 8] [22, 25], [50, 80]
> 
> Then we check the sstable against the trimed cache.
> 
> range [1, 8] hits tokens within the sstable, we invalidate cache [1, 8]. range [22, 25] does not hit anyt oken within the sstable, we skip to invalidate cache [22, 25] range [50, 80] hit tokens within sstable, we skip to invalidate cache [50, 80]

There seems to be a misunderstanding. The cost of invalidation itself is negligible (and if you invalidate a range which is empty in the cache, the cost is 0, because there is nothing to iterate over). The problem with invalidating more than necessary is not that the process of invalidation is costly, but that it lowers the cache hit ratio until the invalidated entries are reloaded from disk. So the problem in this issue is that ranges [1, 8] [22, 25], [50, 80] get invalidated, and the idea doesn't help with that at all. It only skips over the gaps ([9,21], [26, 49], [81, 1000], but they don't matter. You might as well invalidate them, because it's free.

**Comment by asias** [comment]
> > Consider:
> > cache [1, 10] [20, 30] [50,100] owed token range [1, 8] [22, 25] [50, 80] sstable range [1, 1000], which only has 3 tokens 1 , 70, 1000
> > with intersection idea, we first drop the cache so that we have the new cache
> > cache [1, 8] [22, 25], [50, 80]
> > Then we check the sstable against the trimed cache.
> > range [1, 8] hits tokens within the sstable, we invalidate cache [1, 8]. range [22, 25] does not hit anyt oken within the sstable, we skip to invalidate cache [22, 25] range [50, 80] hit tokens within sstable, we skip to invalidate cache [50, 80]
> 
> There seems to be a misunderstanding. The cost of invalidation itself is negligible (and if you invalidate a range which is empty in the cache, the cost is 0, because there is nothing to iterate over). The problem with invalidating more than necessary is not that the process of invalidation is costly, but that it lowers the cache hit ratio until the invalidated entries are reloaded from disk. So the problem in this issue is that ranges [1, 8] [22, 25], [50, 80] get invalidated, and the idea doesn't help with that at all. It only skips over the gaps ([9,21], [26, 49], [81, 1000], but they don't matter. You might as well invalidate them, because it's free.

Yes, I am aware that the motivation is to skip unnecessary invalidation to keep the cache hot if possible. In my example above, we do not invalidate the cache of range [22, 25]. That is the win. If we invalidate from [1, 1000], [22,25] will be invalidated.

**Comment by michoecho** [comment]
Ah, I see now. You said `3. The token range of the sstable being added` so I thought that you wanted to stick to using [`sstable_min_token`, `sstable_max_token`], but understand that you actually want to use the sstable index to find for each owned range the maximal subrange of the sstable which is fully contained in the owned range, and invalidate these subranges?

Sound good.

**Comment by mykaul** [comment]
@asias - did you get a chance to work on this?

**Comment by avikivity** [comment]
We can ignore range queries for now. They make a small proportion of queries (especially since they return large pages).

**Comment by mykaul** [comment]
@denesb - let's evaluate this - priority, complexity.

**Comment by denesb** [comment]
To summarize:
* in the case of classic streaming, the worst case is that the entire cache is invalidated
* in the case of RBNO and repair, the worst case is that all repaired vnodes are invalidated -- in the case of a full repair, this can be almost as bad as in the case of streaming, but it happens gradually.

This issue remains important with tablets, especially in the context of automatic and incremental repair, as this issue can provide performance ripple effects to an otherwise quick repair.

As for the solution, I think this would be the best:
> invalidate by partition. Walk the cache partitions, check if the partition exists in the repair sstables using bloom filters, invalidate if there is a hit. This will work well except for large partitions.

I don't see the problem with large partitions, I think this will work well in all cases. It will cheaper, the smaller the repair will be, which is the direction we want to go towards anyway.

I think this is 1-2 day of coding + testing.

**Comment by dani-tweig** [comment]
> I think this is 1-2 day of coding + testing.

@denesb , how important is it? 
Will it make to 2025.2? (if not please update the milestone)

**Comment by denesb** [comment]
This is an important issue, it is scheduled to be done in Q2.

**Comment by dani-tweig** [comment]
@denesb , @Deexie , will this make it to 2025.3 (Jly 1st) or to 2025.4 (Oct 1st)?

**Comment by denesb** [comment]
> [@denesb](https://github.com/denesb) , [@Deexie](https://github.com/Deexie) , will this make it to 2025.3 (Jly 1st) or to 2025.4 (Oct 1st)?

We haven't even started working on this, because the RF 0->N work was prioritized over this, so 2025.4 at best.
